src/dividat-driver/senso/update_firmware.go

```diff
@@ -0,0 +1,66 @@
+package senso
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/dividat/driver/src/dividat-driver/firmware"
+)
+
+type OnUpdate func(msg FirmwareUpdateMessage)
+
+func (handle *Handle) isUpdatingFirmware() bool {
+	handle.firmwareUpdateMutex.Lock()
+	state := handle.firmwareUpdateInProgress
+	handle.firmwareUpdateMutex.Unlock()
+	return state
+}
+
+func (handle *Handle) setUpdatingFirmware(state bool) {
+	handle.firmwareUpdateMutex.Lock()
+	handle.firmwareUpdateInProgress = state
+	handle.firmwareUpdateMutex.Unlock()
+}
+
+// Disconnect from current connection
+func (handle *Handle) ProcessFirmwareUpdateRequest(command UpdateFirmware, onUpdate OnUpdate) {
+	handle.log.Info("Processing firmware update request.")
+	handle.setUpdatingFirmware(true)
+	if handle.cancelCurrentConnection != nil {
+		msg := "Disconnecting from the Senso"
+		onUpdate(FirmwareUpdateMessage{FirmwareUpdateProgress: &msg})
+		handle.cancelCurrentConnection()
+	}
+	image, err := decodeImage(command.Image)
+	if err != nil {
+		msg := fmt.Sprintf("Error decoding base64 string: %v", err)
+		onUpdate(FirmwareUpdateMessage{FirmwareUpdateFailure: &msg})
+		handle.log.Error(msg)
+	}
+
+	onProgress := func(progressMsg string) {
+		onUpdate(FirmwareUpdateMessage{FirmwareUpdateProgress: &progressMsg})
+	}
+
+	err = firmware.Update(context.Background(), image, nil, &command.Address, onProgress)
```

Comment:
A short wait before triggering the Update routine should help make sure TCP connection is torn down on both ends.

Reply:
I added a 10 second timeout in [Wait 10 seconds for tcp connection teardown](https://github.com/dividat/driver/pull/127/commits/94b819a57baf8bb70375236b5cba8590852c5102), but later  found this to be unnecessary, especially with [Send DFU with exponential backoff](https://github.com/dividat/driver/pull/127/commits/03b962d74561ee234be5e016be6fc3df5b698bd2).